### PR TITLE
Use an installation UUID as device identifier.

### PR DESCRIPTION
**Pull request by orospakr**

### CODE REVIEW

File: rover/src/main/kotlin/io/rover/rover/platform/DeviceIdentification.kt [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    private val storageContextIdentifier = "io.rover.rover.device-identification"
+    private val identifierKey = "identifier"
+    private val storage = localStorage.getKeyValueStorageFor(storageContextIdentifier)
+
+    override val installationIdentifier by lazy {
+        // Further reading: https://developer.android.com/training/articles/user-data-ids.html
+
+        // if persisted UUID not present then generate and persist a new one. Memoize it in memory.
+        storage.get(identifierKey) ?: (
+            UUID.randomUUID().toString().apply {
```
Comment (seanrucker) [inline rover/src/main/kotlin/io/rover/rover/platform/DeviceIdentification.kt:24]: How did you decide between using a randomly generated UUID and the InstanceID service provided by Google Play Services? Google seems to be suggesting the Instance ID unless it isn't "practical". 

> In cases where an Instance ID isn't practical, custom globally unique IDs (GUIDs) can also be used

File: rover/src/main/kotlin/io/rover/rover/platform/LocalStorage.kt [hunk 2]
```diff
@@ -0,0 +1,39 @@
+package io.rover.rover.platform
+
+import android.content.Context
+import android.content.Context.MODE_PRIVATE
+
+/**
+ * Very simple hash-like storage of keys and values.
+ */
+interface KeyValueStorage {
```
Comment (seanrucker) [inline rover/src/main/kotlin/io/rover/rover/platform/LocalStorage.kt:9]: Do you need both the LocalStorage and KeyValueStorage interfaces? Could they be merged?

File: rover/src/main/kotlin/io/rover/rover/platform/DeviceIdentification.kt (hunk 1 above)
Comment (orospakr) [reply rover/src/main/kotlin/io/rover/rover/platform/DeviceIdentification.kt:24]: InstanceID is from Google's Play Services, which I understand we don't want to depend on in the Base SDK.

Comment (seanrucker) [reply rover/src/main/kotlin/io/rover/rover/platform/DeviceIdentification.kt:24]: Makes sense. I didn't know we could get away without depending on Play Services, but if that is an option it is ideal!

File: rover/src/main/kotlin/io/rover/rover/platform/LocalStorage.kt (hunk 2 above)
Comment (orospakr) [reply rover/src/main/kotlin/io/rover/rover/platform/LocalStorage.kt:9]: On Android you can create as many shared preferences instances as you like (and they're each given an identifier).  This interface is meant to be a facade around that.  This means different Rover services can have their own separate storage area, and avoid key collisions between them.

Merging them together is fine if we would prefer to have a single shared preferences amongst the Rover services (which would probably be fine too).

Comment (seanrucker) [reply rover/src/main/kotlin/io/rover/rover/platform/LocalStorage.kt:9]: I don't think we need to have separate shared preferences between services, can't think of a use-case for that. Your call if you want to merge these or keep them separate.

Comment (orospakr) [reply rover/src/main/kotlin/io/rover/rover/platform/LocalStorage.kt:9]: hmm, Having them all together does make clear/reset/logout a little simpler.

**Review by seanrucker — APPROVED** [review]

**Pull request merged**